### PR TITLE
LPS-143549 Modify the logic so that we always call

### PR DESCRIPTION
**Pull request by mbowerman**
Hey @drewbrokke,

This wasn't one of the fixes you suggested, but it should be fairly straightforward and should be regression-proof. Basically the change in logic is this:

**Before This Change**
No Orgs/UserGroups: `UserLocalServiceUtil.getGroupPrimaryKeys(userId)`
Else: `GroupLocalServiceUtil.getUserGroups`

**After This Change**
No Orgs/UserGroups: `UserLocalServiceUtil.getGroupPrimaryKeys(userId)`
Else: `GroupLocalServiceUtil.getUserGroups UNION UserLocalServiceUtil.getGroupPrimaryKeys(userId)`

So the only difference is now in the else case, where the user has at least 1 org or user group, we are UNIONING the result of the else case with the result of the No Orgs/UserGroups case.

Let me know what you think of this.

### CODE REVIEW

**Review by drewbrokke — CHANGES_REQUESTED** [review]
This makes sense, thanks @mbowerman. The last thing that's needed for this is an integration test case. You could either add it in the depot related modules, or you could come up with a generic example and put it in `UserBagFactoryTest`, up to you.

**Comment by mbowerman** [comment]
@drewbrokke I added a commit that adds an integration test.

File: modules/apps/depot/depot-test/src/testIntegration/java/com/liferay/depot/internal/search/test/DepotEntrySearchTest.java [hunk 1]
```diff
@@ -202,4 +276,10 @@ private DepotEntry _addDepotEntry(User user, String name) throws Exception {
 	@Inject
 	private DepotEntryService _depotEntryService;
 
+	@DeleteAfterTestRun
+	private Organization _organization;
```
Comment (pei-jung) [inline modules/apps/depot/depot-test/src/testIntegration/java/com/liferay/depot/internal/search/test/DepotEntrySearchTest.java:280]: This is not needed. DataGuard will automatically clean up the organization after the test.

File: modules/apps/depot/depot-test/src/testIntegration/java/com/liferay/depot/internal/search/test/DepotEntrySearchTest.java [hunk 2]
```diff
@@ -180,6 +187,73 @@ public void testSearchOneWithPermissions() throws Exception {
...
+					birthdayCal.get(Calendar.MONTH),
+					birthdayCal.get(Calendar.DATE),
+					birthdayCal.get(Calendar.YEAR), contact.getSmsSn(),
+					contact.getFacebookSn(), contact.getJabberSn(),
+					contact.getSkypeSn(), contact.getTwitterSn(),
+					user.getJobTitle(), new long[] {depotEntry.getGroupId()},
+					new long[] {_organization.getOrganizationId()}, null, null,
+					user.getUserGroupIds(),
+					ServiceContextTestUtil.getServiceContext(
+						TestPropsValues.getGroupId(), adminUser.getUserId()));
```
Comment (pei-jung) [inline modules/apps/depot/depot-test/src/testIntegration/java/com/liferay/depot/internal/search/test/DepotEntrySearchTest.java:228]: I think we can simplify the code by using `DepotTestUtil.withAssetLibraryMember`. That way we end up with a user that is a member of the depot entry. Then, we can assign the user to the organization via `_userService.addOrganizationUsers`.

**Review by pei-jung — CHANGES_REQUESTED** [review]

**Comment by mbowerman** [comment]
@pei-jung Thanks for the feedback. I've made the requested changes and force-pushed the commit. I also re-ran the test locally to verify that it still passes.

I do have one question though. Isn't it also true that the DepotEntries will also be deleted by the DataGuard after the test? If that's the case, shouldn't we remove the annotation from [here](https://github.com/liferay-usm/liferay-portal/pull/390/commits/be59872dc79048c1b4d0d3b28f31003b47f96781#diff-74230f64f5d96242866d7a54a02d5f31e54588cb264a119c3f025ca7281f64deL199-L200) too?

**Comment by pei-jung** [comment]
@mbowerman yes, the `_depotEntries` can also be removed if its purpose is only for test cleanup. DataGuard was introduced about a year or two ago, so it's relatively new. Some of the existing tests have not been refactored yet to take advantage of this new test rule.

**Comment by mbowerman** [comment]
Hey @pei-jung I tried removing the `_depotEntries`, but then when I reran the test class, I got a test failure at `testSearchOneWithPermissions`:

```
com.liferay.portal.kernel.exception.DuplicateGroupException: {groupId=43438}
        at com.liferay.portal.service.impl.GroupLocalServiceImpl.validateGroupKey(GroupLocalServiceImpl.java:4962)
        at com.liferay.portal.service.impl.GroupLocalServiceImpl.addGroup(GroupLocalServiceImpl.java:436)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50)
        at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.service.ServiceContextAdvice.invoke(ServiceContextAdvice.java:60)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49)
        at com.sun.proxy.$Proxy73.addGroup(Unknown Source)
        at com.liferay.depot.service.impl.DepotEntryLocalServiceImpl.addDepotEntry(DepotEntryLocalServiceImpl.java:116)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50)
        at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.service.ServiceContextAdvice.invoke(ServiceContextAdvice.java:60)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.change.tracking.internal.aop.CTTransactionAdvice.invoke(CTTransactionAdvice.java:80)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49)
        at com.sun.proxy.$Proxy580.addDepotEntry(Unknown Source)
        at com.liferay.depot.service.impl.DepotEntryServiceImpl.addDepotEntry(DepotEntryServiceImpl.java:62)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50)
        at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.service.ServiceContextAdvice.invoke(ServiceContextAdvice.java:60)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.kernel.aop.ChainableMethodAdvice.invoke(ChainableMethodAdvice.java:64)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.change.tracking.internal.aop.CTTransactionAdvice.invoke(CTTransactionAdvice.java:80)
        at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
        at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49)
        at com.sun.proxy.$Proxy583.addDepotEntry(Unknown Source)
        at com.liferay.depot.internal.search.test.DepotEntrySearchTest._addDepotEntry(DepotEntrySearchTest.java:236)
        at com.liferay.depot.internal.search.test.DepotEntrySearchTest.testSearchOneWithPermissions(DepotEntrySearchTest.java:145)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$3.evaluate(TestExecutorRunnable.java:353)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$1.evaluate(TestExecutorRunnable.java:237)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable._execute(TestExecutorRunnable.java:316)
        at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable.run(TestExecutorRunnable.java:95)
        at java.lang.Thread.run(Thread.java:748)
```

Maybe there is a difference between DataGuard and @DeleteAfterTestRun? Does @DeleteAfterTestRun do the cleanup after individual test methods, whereas DataGuard only does the cleanup after then entire test class? That could explain why I saw this error after removing the `_depotEntries`.

**Comment by pei-jung** [comment]
hi @mbowerman, your assessment is correct. DataGuard by default does the cleanup after the entire test class. If you want to do the cleanup after each test methods, you can change the scope by doing `@DataGuard(scope = DataGuard.Scope.METHOD)`. Just add the annotation to the test class.

**Comment by mbowerman** [comment]
Thanks @pei-jung. I did as you suggested and the test is passing again. I updated the pull request with a new commit.

**Review by pei-jung — APPROVED** [review]
Changes look good. Thanks @mbowerman!

**Comment by pei-jung** [comment]
ci:forward

**Pull request closed without merging**